[SDXL and IP2P]: instruction pix2pix XL training and pipeline

harutatsuakiyama opened the pull request:
### Support for Training Instruct Pix2Pix with Stable Diffusion XL
**Description**:

This pull request introduces support for training Instruct Pix2Pix with Stable Diffusion XL. The changes include two new files:

train_instruct_pix2pix_xl.py: This file contains the necessary code to train the Instruct Pix2Pix model using the Stable Diffusion XL method. It includes the training loop, loss function calculations, and backpropagation steps.

pipeline_stable_diffusion_xl_instruct_pix2pix.py: This file provides the pipeline for inferencing with the trained Instruct Pix2Pix model. It handles the pre-processing of input data, model inference, and post-processing of model outputs.

The addition of these files will allow users to train and inference with the Instruct Pix2Pix model of Stable Diffusion XL, which can potentially lead to improved performance and results.

harutatsuakiyama commented:
Bugs observed. Fixing now.

andreemic commented:
Hey what bugs did you observe? I was working on the same thing, maybe I can help.

harutatsuakiyama commented:
> Hey what bugs did you observe? I was working on the same thing, maybe I can help.

Training generates weird artifacts like below.
![edited_image_xl](https://github.com/huggingface/diffusers/assets/8084808/9ed7038f-7ea7-4af1-b436-82dd6e352b28)
Looking into this now.

harutatsuakiyama commented:
> > Hey what bugs did you observe? I was working on the same thing, maybe I can help.
> 
> Training generates weird artifacts like below. ![edited_image_xl](https://user-images.githubusercontent.com/8084808/253740262-9ed7038f-7ea7-4af1-b436-82dd6e352b28.png) Looking into this now.

Now bugs have been fixed. Reasonable images can be generated. Working on cleaning up code now.

andreemic commented:
Did you take an SDXL pipeline as base? Or what did you start with?

harutatsuakiyama commented:
> Did you take an SDXL training pipeline as base? Or what did you start with?

I modified the original instruct pix2pix training pipeline to be compatible with SDXL. You may refer to the code.

harutatsuakiyama commented:
Huggingface test says this part has error: 

```
examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:19:1: I001 [*] Import block is un-sorted or un-formatted
examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:49:53: F401 [*] `diffusers.StableDiffusionInstructPix2PixPipeline` imported but unused
Found 2 errors.
```

I have difficulties to understand. If possible, can people give some hints?

sayakpaul commented:
@harutatsuakiyama oh wow! InstructPix2Pix is always very dear to my heart. This is why I worked on this: https://huggingface.co/blog/instruction-tuning-sd. 

But anyway, could I get some results from your training runs before I do a deep review of your PR if it's not too much?

harutatsuakiyama commented:
> @harutatsuakiyama oh wow! InstructPix2Pix is always very dear to my heart. This is why I worked on this: https://huggingface.co/blog/instruction-tuning-sd.
> 
> But anyway, could I get some results from your training runs before I do a deep review of your PR if it's not too much?

Sure. I trained on this dataset: https://www.zheyuanliu.me/CIRR/, since I have trained this dataset with IP2P before. 

These are some examples: 
1. 
_Has dining room view, Change stools to long white benches, Add a chair on the right corner_
![reference_0](https://github.com/huggingface/diffusers/assets/8084808/2dd1185c-d2f8-4a5d-aaf0-7091574e428e)
![edited_image_xl_new_1](https://github.com/huggingface/diffusers/assets/8084808/4ee2b080-5a15-418f-9127-23ced50fdd78)

2. 
_Has similar layout of bed and pillows, Change bed to pink_
![reference_3](https://github.com/huggingface/diffusers/assets/8084808/1366e46c-45a8-455b-b9fd-716bff2aa087)
![edited_image_xl_new_3_2](https://github.com/huggingface/diffusers/assets/8084808/4942a90c-ab68-480f-8e74-a6aeafbfc68e)

The dataset is not very pix2pix, hence the change is big.

harutatsuakiyama commented:
@sayakpaul Let me know if you need further information.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1113 @@
+"""Script to train Stable Diffusion XL for InstructPix2Pix."""
```
We usually added licensing here instead of this. I believe this could go to the description of the argument parser.

harutatsuakiyama replied:
Added licensing. Put this in the description of the argument parser.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1113 @@
+
+logger = get_logger(__name__, log_level="INFO")
+
+DATASET_NAME_MAPPING = {
+    "fusing/instructpix2pix-1000-samples": ("file_name", "edited_image", "edit_prompt"),
+}
```
Is your dataset private? If not, let's maybe include it here?

harutatsuakiyama replied:
You mean CIRR dataset? We can put it publicly on HF later. I think it is not top priority now :-)

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1113 @@
+        weight_dtype = torch.bfloat16
+        warnings.warn(f"weight_dtype {weight_dtype} may cause nan during vae encoding", UserWarning)
+
+    # Move text_encode and vae to gpu and cast to weight_dtype
+    text_encoder_1.to(accelerator.device, dtype=weight_dtype)
+    text_encoder_2.to(accelerator.device, dtype=weight_dtype)
+    text_encoders = [text_encoder_1, text_encoder_2]
+    vae.to(accelerator.device, dtype=weight_dtype)
```
Since the text encoders are not trained, let's precompute all the text embeddings and the other conditional embeddings. What say? 

This will help us a lot of a memory.

harutatsuakiyama replied:
Made it precomputed.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1113 @@
+                    # Final text conditioning.
+                    ### Begin: Get null conditioning
+                    null_conditioning_list = []
+                    for a_tokenizer, a_text_encoder in zip(
+                        (tokenizer_1, tokenizer_2), (text_encoder_1, text_encoder_2)
+                    ):
+                        null_conditioning_list.append(
+                            a_text_encoder(
+                                tokenize_captions([""], a_tokenizer=a_tokenizer).to(accelerator.device),
+                                output_hidden_states=True,
+                            ).hidden_states[-2]
```
Similarly we can prepcompute this.

harutatsuakiyama replied:
Made it precomputed.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1113 @@
+                    raise ValueError(f"Unknown prediction type {noise_scheduler.config.prediction_type}")
+
+                ### Begin SDXL
+                add_text_embeds = pooled_prompt_embeds
+
+                crops_coords_top_left = (0, 0)
+                target_size = (args.resolution, args.resolution)
+                original_size = original_image_embeds.shape[-2:]
+                add_time_ids = list(original_size + crops_coords_top_left + target_size)
+                add_neg_time_ids = list(original_size + crops_coords_top_left + target_size)
+                add_time_ids = torch.tensor([add_time_ids], dtype=encoder_hidden_states.dtype)
+                add_neg_time_ids = torch.tensor([add_neg_time_ids], dtype=encoder_hidden_states.dtype)
+                add_time_ids = add_time_ids.to(encoder_hidden_states.device).repeat(args.train_batch_size, 1)
+                ### End SDXL
+
+                # Predict the noise residual and compute loss
+                added_cond_kwargs = {"text_embeds": add_text_embeds, "time_ids": add_time_ids}
```
Let's precompute. Reference is available here: 

https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_sdxl.py

Let's also accept `crops_coords_top_left` as CLI args as done in the above script.

harutatsuakiyama replied:
Made it precomputed.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1113 @@
+
+                crops_coords_top_left = (0, 0)
+                target_size = (args.resolution, args.resolution)
+                original_size = original_image_embeds.shape[-2:]
```
Why are we determining the original image resolution from the image embeddings?

harutatsuakiyama replied:
This might be a bug. Fixed already.

sayakpaul reviewed (commented):
Cool, thank you!

This is looking good. I left a couple of comments. 

Some follow-up questions:

* Have you tried using this improved VAE https://huggingface.co/madebyollin/sdxl-vae-fp16-fix? 
* Did you face any numerical instability issues in FP16 when using the default VAE shipped with SDXL? 
* Could you leave a link the `wandb` training run page so that we could see the dynamics of the training? 
* Did you notice any qualitative difference in between SDXL and SD InstructPix2Pix on your dataset? 

After the comments are addressed, we need to add:

- [ ] Docs
- [ ] Tests
- [ ] Misc changes (such as inits, dummy objects, etc.) 

Let me know in case anything is unclear.

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_instruct_pix2pix.py:
```diff
@@ -0,0 +1,943 @@
+        """
+        self.vae.disable_tiling()
+
+    def enable_sequential_cpu_offload(self, gpu_id=0):
+        r"""
+        Offloads all models to CPU using accelerate, significantly reducing memory usage. When called, unet,
+        text_encoder, vae and safety checker have their state dicts saved to CPU and then are moved to a
+        `torch.device('meta') and loaded to GPU only when their specific submodule has its `forward` method called.
+        Note that offloading happens on a submodule basis. Memory savings are higher than with
+        `enable_model_cpu_offload`, but performance is lower.
+        """
+        if is_accelerate_available() and is_accelerate_version(">=", "0.14.0"):
+            from accelerate import cpu_offload
+        else:
+            raise ImportError("`enable_sequential_cpu_offload` requires `accelerate v0.14.0` or higher")
+
+        device = torch.device(f"cuda:{gpu_id}")
+
+        if self.device.type != "cpu":
+            self.to("cpu", silence_dtype_warnings=True)
+            torch.cuda.empty_cache()  # otherwise we don't see the memory savings (but they probably exist)
+
+        for cpu_offloaded_model in [self.unet, self.text_encoder_2, self.vae]:
+            cpu_offload(cpu_offloaded_model, device)
+
+        if self.text_encoder is not None:
+            cpu_offload(self.text_encoder, device)
+
```
```suggestion
```

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_instruct_pix2pix.py:
```diff
@@ -0,0 +1,943 @@
+
+        # We'll offload the last model manually.
+        self.final_offload_hook = hook
+
+    @property
+    def _execution_device(self):
+        r"""
+        Returns the device on which the pipeline's models will be executed. After calling
+        `pipeline.enable_sequential_cpu_offload()` the execution device can only be inferred from Accelerate's module
+        hooks.
+        """
+        if not hasattr(self.unet, "_hf_hook"):
+            return self.device
+        for module in self.unet.modules():
+            if (
+                hasattr(module, "_hf_hook")
+                and hasattr(module._hf_hook, "execution_device")
+                and module._hf_hook.execution_device is not None
+            ):
+                return torch.device(module._hf_hook.execution_device)
+        return self.device
```
```suggestion
```

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_instruct_pix2pix.py:
```diff
@@ -0,0 +1,943 @@
+                    if callback is not None and i % callback_steps == 0:
+                        callback(i, t, latents)
+
+        # make sure the VAE is in float32 mode, as it overflows in float16
+        self.vae.to(dtype=torch.float32)
+
+        use_torch_2_0_or_xformers = isinstance(
+            self.vae.decoder.mid_block.attentions[0].processor,
+            (
+                AttnProcessor2_0,
+                XFormersAttnProcessor,
+                LoRAXFormersAttnProcessor,
+                LoRAAttnProcessor2_0,
+            ),
+        )
+        # if xformers or torch_2_0 is used attention block does not need
+        # to be in float32 which can save lots of memory
+        if use_torch_2_0_or_xformers:
+            self.vae.post_quant_conv.to(latents.dtype)
+            self.vae.decoder.conv_in.to(latents.dtype)
+            self.vae.decoder.mid_block.to(latents.dtype)
+        else:
+            latents = latents.float()
```
```suggestion
        # make sure the VAE is in float32 mode, as it overflows in float16
        self.vae.to(dtype=torch.float32)

        use_torch_2_0_or_xformers = isinstance(
            self.vae.decoder.mid_block.attentions[0].processor,
            (
                AttnProcessor2_0,
                XFormersAttnProcessor,
                LoRAXFormersAttnProcessor,
                LoRAAttnProcessor2_0,
            ),
        )
        # if xformers or torch_2_0 is used attention block does not need
        # to be in float32 which can save lots of memory
        if use_torch_2_0_or_xformers:
            self.vae.post_quant_conv.to(latents.dtype)
            self.vae.decoder.conv_in.to(latents.dtype)
            self.vae.decoder.mid_block.to(latents.dtype)
        else:
            latents = latents.float()
```

Could we try to align this more with the new design: https://github.com/huggingface/diffusers/blob/ed2a3584ab2ee59676eb95c884ba2d7bb831f41d/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L790

patrickvonplaten commented:
Could we also add some tests for the new pipeline class? Looks very cool!

sayakpaul commented:
> Could we also add some tests for the new pipeline class? Looks very cool!

TODOs are well laid out in https://github.com/huggingface/diffusers/pull/4079#pullrequestreview-1536181349 :-) 

We're happy to help with anything to ship this 🚀

harutatsuakiyama commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1113 @@
+        result = urlparse(string)
+        return all([result.scheme, result.netloc])
+    except ValueError:
+        return False
```
Made it as a closure.

sayakpaul commented:
Thanks for the changes @harutatsuakiyama! Let us know when you'd like another round of feedback.

harutatsuakiyama commented:
Hi @sayakpaul . Thank you for great suggestions! I have addressed them in the new commit a4772e7. 

For the following up questions: 
_Have you tried using this improved VAE https://huggingface.co/madebyollin/sdxl-vae-fp16-fix?_
No. I did get nan when using fp16 for VAE. Will try later in a separate PR. 

_Did you face any numerical instability issues in FP16 when using the default VAE shipped with SDXL?_
Yes. during inference using the firstly trained IP2P SDXL, I got nan errors. It took me some time to figure out the problem lied in the VAE. 

_Could you leave a link the wandb training run page so that we could see the dynamics of the training?_
I didn't use the wandb. Will use in the future. 

_Did you notice any qualitative difference in between SDXL and SD InstructPix2Pix on your dataset?_
I couldn't find the previously trained IP2P model on CIRR. Need to retrain that ...

harutatsuakiyama commented:
> Thanks for the changes @harutatsuakiyama! Let us know when you'd like another round of feedback.

@sayakpaul I will address other requests. In the meantime, can you please have a look at the updated train_instruct_pix2pix_xl.py?

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1200 @@
+WANDB_TABLE_COL_NAMES = ["file_name", "edited_image", "edit_prompt"]
+
+
+# Load image from URL or path
```
I don't think we need to do a closure like this for a simple utility like this. This can stay in the main() fun as this way the code becomes more readable. Or if we were to go with a standalone function we can just make it a standalone one and not one that has a closure.

harutatsuakiyama replied:
Changed the code to have just a single line doing this end.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1200 @@
+        default=256,
+        help=(
+            "The resolution for input images, all the images in the train/validation dataset will be resized to this"
+            " resolution."
```
Can be clubbed with the previous line no?

harutatsuakiyama replied:
Done.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1200 @@
+        project_config=accelerator_project_config,
+    )
+
+    generator = torch.Generator(device=accelerator.device).manual_seed(args.seed)
```
Why are we instantiating a generator here where we know that it's use might be conditional.

harutatsuakiyama replied:
I followed the original train_iinstruct_pix2pix code. They have a generator.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1200 @@
+        prompt_embeds = torch.concat([example["prompt_embeds"] for example in examples], dim=0)
+        add_text_embeds = torch.concat([example["add_text_embeds"] for example in examples], dim=0)
+        return {
+            "original_pixel_values": original_pixel_values,
```
Why are we not including the time ids here?

harutatsuakiyama replied:
time ids I think is a constant variable. We can just create a single one and use it each time. No need to duplicate it in my opinion.

sayakpaul reviewed (commented):
Thanks for iterating. Just left some comments on the training script.

harutatsuakiyama commented:
@sayakpaul Thank you for the new comments. Will address them in this weekend. Busy with other work now.

harutatsuakiyama commented:
@patrickvonplaten Thank you for suggestions :-) I have adapted the pipeline with the diffusers/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py. Test has also been created at tests.pipelines.stable_diffusion_xl.test_stable_diffusion_xl_instruction_pix2pix.py. 

@sayakpaul Thank you for the new comments :-) I have replied and addressed them.

For docs, I don't know why after I have changed the docs, it can't pass the test. Can you help with that? I think basically, just add a small part in the original IP2P doc about how to train with XL.

harutatsuakiyama commented:
Ohhh test errors. Let me have a look.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1185 @@
+
+    # Preprocessing the datasets.
+    # We need to tokenize input captions and transform the images.
+    def tokenize_captions(captions, a_tokenizer):
```
```suggestion
    def tokenize_captions(captions, tokenizer):
```

harutatsuakiyama replied:
Changed.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1185 @@
+    null_conditioning = compute_null_conditioning()
+
+    def compute_time_ids():
+        crops_coords_top_left = (0, 0)
```
Let's make this CLI too. Example: https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_sdxl.py

harutatsuakiyama replied:
Made them as CLI.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1185 @@
+                continue
+
+            with accelerator.accumulate(unet):
+                # We want to learn the denoising process w.r.t the edited images which
+                # are conditioned on the original image (which was edited) and the edit instruction.
+                # So, first, convert images to latent space.
+                latents = vae.encode(batch["edited_pixel_values"].to(weight_dtype)).latent_dist.sample()
+                latents = latents * vae.config.scaling_factor
```
Could we follow the method of encoding and casting followed in https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_sdxl.py?

harutatsuakiyama replied:
Implemented.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1185 @@
+
+                # Get the additional image embedding for conditioning.
+                # Instead of getting a diagonal Gaussian here, we simply take the mode.
+                original_image_embeds = vae.encode(batch["original_pixel_values"].to(weight_dtype)).latent_dist.mode()
```
Same: https://github.com/huggingface/diffusers/pull/4079/files#r1270503918

harutatsuakiyama replied:
Implemented.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1185 @@
+                        ema_unet.copy_to(unet.parameters())
+
+                    # The models need unwrapping because for compatibility in distributed training mode.
+                    pipeline = StableDiffusionXLInstructPix2PixPipeline.from_pretrained(
+                        args.pretrained_model_name_or_path,
+                        unet=accelerator.unwrap_model(unet),
+                        text_encoder=accelerator.unwrap_model(text_encoder_1),
+                        text_encoder_2=accelerator.unwrap_model(text_encoder_2),
+                        tokenizer=accelerator.unwrap_model(tokenizer_1),
+                        tokenizer_2=accelerator.unwrap_model(tokenizer_2),
+                        vae=accelerator.unwrap_model(vae),
+                        revision=args.revision,
+                        torch_dtype=weight_dtype,
+                    )
```
If we're precomputing the text embeddings, we should free the memory after it's done. You can find a reference in https://github.com/huggingface/diffusers/pull/4079/files#r1270503918. 

In this case, we should only unwrap the UNet. VAE can be reused as is.

harutatsuakiyama replied:
Changed.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1185 @@
+        if args.use_ema:
+            ema_unet.copy_to(unet.parameters())
+
+        pipeline = StableDiffusionXLInstructPix2PixPipeline.from_pretrained(
+            args.pretrained_model_name_or_path,
+            text_encoder=accelerator.unwrap_model(text_encoder_1),
+            text_encoder_2=accelerator.unwrap_model(text_encoder_2),
+            tokenizer=accelerator.unwrap_model(tokenizer_1),
+            tokenizer_2=accelerator.unwrap_model(tokenizer_2),
+            vae=accelerator.unwrap_model(vae),
+            unet=unet,
+            revision=args.revision,
+        )
```
Same.

harutatsuakiyama replied:
Changed.

sayakpaul reviewed (commented):
The training code changes are looking good. Let's also make sure to add some documentation:

- A separate README (README_sdxl.md) including comments, example commands, etc. 
- Mention that from the original README of InstructPix2Pix training example. 
- Add a test for the training example. 

Let me know if anything is unclear. 

@patrickvonplaten could you review the rest of the updates?

harutatsuakiyama commented:
Hi @sayakpaul @patrickvonplaten , 

I have addressed the code reviews. A README_sdxl.md on how to train train_instruct_pix2pix_xl.py has been added. 

The README mentioned that from the original README of InstructPix2Pix training example.

I have also updated the test for training and inferencing: tests/pipelines/stable_diffusion/test_stable_diffusion_instruction_pix2pix.bak.py

sayakpaul commented on tests/pipelines/stable_diffusion_xl/test_stable_diffusion_xl_instruction_pix2pix.py:
```diff
@@ -0,0 +1,155 @@
+import random
```
Missing licensing.

harutatsuakiyama replied:
Added licence.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_instruct_pix2pix.py:
```diff
@@ -0,0 +1,947 @@
+import inspect
```
Missing licensing.

harutatsuakiyama replied:
Added licence.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1188 @@
+                        os.makedirs(val_save_dir)
+
+                    original_image = (
+                        lambda image_url_or_path: Image.open(BytesIO(requests.get(image_url_or_path).content)).convert(
+                            "RGB"
+                        )
+                        if urlparse(image_url_or_path).scheme
```
We can simply use `load_image()` from `diffusers.utils` when loading an image from a URL.

harutatsuakiyama replied:
Fixed.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1188 @@
+                # We want to learn the denoising process w.r.t the edited images which
+                # are conditioned on the original image (which was edited) and the edit instruction.
+                # So, first, convert images to latent space.
+                latents = vae.encode(batch["edited_pixel_values"]).latent_dist.sample()
+                latents = latents * vae.config.scaling_factor
+                latents = latents.to(weight_dtype)
```
I think this doesn't yet handle the case when the underlying VAE is the original SDXL VAE. See these LOC to check how this should be handled:

https://github.com/huggingface/diffusers/blob/2e53936c97d167713c9e97414160124861fa4b68/examples/controlnet/train_controlnet_sdxl.py#L1119

harutatsuakiyama replied:
Good suggestion. Fixed.

sayakpaul commented on examples/instruct_pix2pix/train_instruct_pix2pix_xl.py:
```diff
@@ -0,0 +1,1188 @@
+
+                # Get the additional image embedding for conditioning.
+                # Instead of getting a diagonal Gaussian here, we simply take the mode.
+                original_image_embeds = vae.encode(batch["original_pixel_values"]).latent_dist.mode()
+                original_image_embeds = original_image_embeds.to(weight_dtype)
```
Same as https://github.com/huggingface/diffusers/pull/4079/files#r1271453644

harutatsuakiyama replied:
Fixed.

sayakpaul commented on examples/instruct_pix2pix/README_sdxl.md:
```diff
@@ -0,0 +1,148 @@
+
+***This is based on the original InstructPix2Pix training example.***
+
+[Stable Diffusion XL](https://github.com/Stability-AI/generative-models/blob/main/assets/sdxl_report.pdf) (or SDXL) is the latest image generation model that is tailored towards more photorealistic outputs with more detailed imagery and composition compared to previous SD models. It leverages a three times larger UNet backbone. The increase of model parameters is mainly due to more attention bloc ...
```
Let's use https://huggingface.co/papers/2307.01952 to cite SDXL :)

harutatsuakiyama replied:
Changed.

sayakpaul commented on examples/instruct_pix2pix/README_sdxl.md:
```diff
@@ -0,0 +1,148 @@
+import PIL
+import requests
+import torch
+from diffusers import StableDiffusionInstructPix2PixPipeline
```
Needs to be StableDiffusionXL pipeline

harutatsuakiyama replied:
Changed.

sayakpaul reviewed (commented):
Thanks for iterating. Left some more comments!

I think the following are the TODOs before we can merge it:

* [ ] Tests for the example (should go in `examples/test_examples.py`)
* [ ] Documentation for the SDXL InstructPix2Pix pipeline. @patrickvonplaten should we update the existing [documentation](https://huggingface.co/docs/diffusers/main/en/api/pipelines/pix2pix) for InstructPix2Pix for this? 
* [ ] Add other misc files needed to package SDXL InstructPix2Pix pipeline in `diffusers`. You can get an idea of how to do that here: https://github.com/huggingface/diffusers/pull/2721. 
* [ ] Discuss your awesome interior design results from the README of the SDXL InstructPix2Pix training briefly. I think the community will feel inspired to see it. 

Let me know if anything is unclear.

blldd commented:
Thanks a lot for your great work, it is very helpful to me, but I wonder why you are training iP2P based on sd-xl-base model instead of sd-xl-refiner, I think refiner may be more suitable for iP2P.

sayakpaul commented:
The base model is more suitable here I think given the training methodology of InstructPix2Pix.

blldd commented:
In addition, during the training process, I encountered the problem of loss isnan error in fp16, and also tried the vae_fix model, and it was also isnan at step 100.

harutatsuakiyama commented:
Hi @sayakpaul and @patrickvonplaten, 

I have addressed the code review. Once I have run more experiments, I will give more detailed documentation on training IP2P with SDXL. 

I will become very busy starting this week, since the semester has started and I need to teach and supervise students. If possible, can we merge this PR? If we have more interesting thoughts, we can address them in the future when I am not so busy.

sayakpaul commented:
> I have addressed the code review. Once I have run more experiments, I will give more detailed documentation on training IP2P with SDXL.

Having some good results documented (like the one you presented on interior design), preferably with the trained checkpoints, will be very beneficial to the community. 

Maybe we could simply write something like WIP on the corresponding pages. @patrickvonplaten WDYT?

harutatsuakiyama commented:
> 

Sure, we can do this in another PR, maybe this weekend. Let's please complete this one now.

patrickvonplaten commented:
> ew. Once I have run more experiments, I will give more detailed documentation on training IP2P with SDXL.

Ok for me!

patrickvonplaten commented:
If ok for @sayakpaul let's merge it :-)

sayakpaul reviewed (approved):
Let's merge. I think it's safe to revisit some stuff (documentation, examples, etc.) later. At least with the PR, we can enable the community.

sayakpaul commented:
Thanks for your contribution!

blldd commented:
> > @harutatsuakiyama oh wow! InstructPix2Pix is always very dear to my heart. This is why I worked on this: https://huggingface.co/blog/instruction-tuning-sd.
> > But anyway, could I get some results from your training runs before I do a deep review of your PR if it's not too much?
> 
> Sure. I trained on this dataset: https://www.zheyuanliu.me/CIRR/, since I have trained this dataset with IP2P before.
> 
> These are some examples:
> 
> 
> _Has dining room view, Change stools to long white benches, Add a chair on the right corner_ ![reference_0](https://user-images.githubusercontent.com/8084808/254413057-2dd1185c-d2f8-4a5d-aaf0-7091574e428e.png) ![edited_image_xl_new_1](https://user-images.githubusercontent.com/8084808/254413092-4ee2b080-5a15-418f-9127-23ced50fdd78.png)
> 
> 
> _Has similar layout of bed and pillows, Change bed to pink_ ![reference_3](https://user-images.githubusercontent.com/8084808/254413646-1366e46c-45a8-455b-b9fd-716bff2aa087.png) ![edited_image_xl_new_3_2](https://user-images.githubusercontent.com/8084808/254413878-4942a90c-ab68-480f-8e74-a6aeafbfc68e.png)
> 
> The dataset is not very pix2pix, hence the change is big.

It's really great, but I follow the training script in reademe, the loss can't converge well, so I wonder if you can share ckpt?

harutatsuakiyama commented:
> > > @harutatsuakiyama oh wow! InstructPix2Pix is always very dear to my heart. This is why I worked on this: https://huggingface.co/blog/instruction-tuning-sd.
> > > But anyway, could I get some results from your training runs before I do a deep review of your PR if it's not too much?
> > 
> > 
> > Sure. I trained on this dataset: https://www.zheyuanliu.me/CIRR/, since I have trained this dataset with IP2P before.
> > These are some examples:
> > _Has dining room view, Change stools to long white benches, Add a chair on the right corner_ ![reference_0](https://user-images.githubusercontent.com/8084808/254413057-2dd1185c-d2f8-4a5d-aaf0-7091574e428e.png) ![edited_image_xl_new_1](https://user-images.githubusercontent.com/8084808/254413092-4ee2b080-5a15-418f-9127-23ced50fdd78.png)
> > _Has similar layout of bed and pillows, Change bed to pink_ ![reference_3](https://user-images.githubusercontent.com/8084808/254413646-1366e46c-45a8-455b-b9fd-716bff2aa087.png) ![edited_image_xl_new_3_2](https://user-images.githubusercontent.com/8084808/254413878-4942a90c-ab68-480f-8e74-a6aeafbfc68e.png)
> > The dataset is not very pix2pix, hence the change is big.
> 
> It's really great, but I follow the training script in reademe, the loss can't converge well, so I wonder if you can share ckpt?

Will give more details on weekends. You will need to increase the batch size to be 4, which will improve performance. Also, be patient. I found loss may oscillate, but generated results are good. 

If you have QQ, you can add me: 290956355. We can discuss more.
